src/lightning/fabric/utilities/distributed.py

```diff
@@ -23,9 +25,62 @@ class group:  # type: ignore
         WORLD = None
 
 
+if TYPE_CHECKING:
+    from lightning.fabric import Fabric
+    from lightning.fabric.strategies import Strategy
+
+
 log = logging.getLogger(__name__)
 
 
+def is_shared_filesystem(fabric: "Fabric", path: Optional[_PATH] = None, timeout: int = 3) -> bool:
+    """Checks whether the filesystem under the given path is shared across all processes.
+
+    Args:
+        fabric: The current :class:`~lightning.fabric.fabric.Fabric` instance
+        path: The path to check. Defaults to the current working directory. The user must have permissions to write
+            to this path or the parent folder, and the filesystem must be writable.
+        timeout: If any of the processes can't list the file created by rank 0 within this timeout in seconds, the
+            filesystem is determined to be not shared.
+
+    """
+    return _is_shared_filesystem(fabric.strategy, path=path, timeout=timeout)
+
+
+def _is_shared_filesystem(strategy: "Strategy", path: Optional[_PATH] = None, timeout: int = 3) -> bool:
```

Comment:
Should this be lru-cached?

Reply:
IMO yes, but we should probably reset the cache on strategy.setup_environment

Reply:
I agree about the cache but I'd prefer not to reach into it like that to reset it. I think doing it anywhere in the Trainer or Fabric is very arbitrary. Let's not overthink it. The expectation is that the filesystem infrastructure doesn't change during the runtime, which is very reasonable to assume IMO.

Reply:
I wasn't worried about the filesystem infrastructure changing but the `world_size` check. For instance, with TPU spawn the `world_size` attribute won't be set until spawn has been launched. So if we cache this before that happens, the output will be incorrect.

Reply:
In this case I would prefer to not have the cache. There is no contract in the strategy that enforces to be aware of this function. I'll add a note to the docstring that the function is meant to be used in a context where distributed is initialized.

Reply:
If we ever do the collectives refactor, we will be able to call into the strategy's collective and check whether it is initialized, so I'm leaving this optimization for the future.